### ISSUE TITLE
qrexec policy domain(s)

### ISSUE THREAD

**Issue by 3hhh** [QubesOS/qubes-issues#7372]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)
qrexec policy decisions are done in dom0.

Also, IIRC the protocol is roughly as following:
1. VM requests a qrexec service from dom0.
2. dom0 makes the policy decision.
3. A _direct_ vchan connection between the two involved VMs is set up, i.e. `[VM A] -----> [VM B]`.
4. dom0 involvement stops.

### The solution you'd like
qrexec policy decisions are done inside one or more dedicated `qrexec` VMs.

Also, I suggest to change the architecture to a proxy model:
1. VM requests a qrexec service from the policy VM that is assigned for the decision (statically configurable for every tuple of VMs).
2. The policy VM makes the decision.
3. The policy VM starts pulling the data from the requesting VM and _actively_ forwards it to the destination VM: `[VM A] -----> policy VM -----> [VM B]`
4. The policy VM stays in control and can actively terminate (e.g. after a configurable timeout) or even monitor connections.

### The value to a user, and who that user might be

dedicated policy VM:
- bugs in qrexec code do not lead to full Qubes compromise anymore
- thinner dom0
- no advanced hypervisor code such as `argo` needed, all can happen in user space

_multiple_ policy VM support (any tuple of VMs can be assigned a policy VM):
- segregate risk more
- Particularly useful for the proxy model if non-vchan (network for Qubes Air) channels are used for qrexec and that network code needs to be put inside a less trusted policy VM.

proxy model:
- precursor for Qubes Air as it's easily adaptable to qrexec over the network (just replace vchan with some encrypted network protocol, consider vchan as a transmission channel like any other)
- more control for the policy VM and thus the user over active qrexec connections; e.g. the following features could be implemented with that model:
  - unidirectional qrexec support (policy VM just proxies in one direction), cf. #6900 
  - do not let malicious VMs keep qrexec channels open forever
  - monitor qrexec channels, allow the user to forcibly close them
  - easier debugging with standard tools (wireshark?)

I'd imagine a very simple policy VM to manage (open & close) [vchan](https://github.com/QubesOS/qubes-core-vchan-socket) or network sockets according to the policy it has and connect them with e.g. `socat` as needed. Some sort of conditional network bridge.

### Drawbacks

- speed & resources (proxying is one memory copy more expensive)
- more danger for security relevant bugs (e.g. if the policy VM enters into some traffic monitoring business or so)

**Comment by 3hhh** [comment]
Btw I think that this is fairly easy to implement with the current architecture and I guess I'll give it a shot once I have some time.
From my perspective it doesn't even require any changes of the existing code.

Any thoughts @DemiMarie @marmarek ?

**Comment by 3hhh** [comment]
> From my perspective it doesn't even require any changes of the existing code.

After some more thinking I believe that some qubes-rpc aliasing policy syntax (translate a qrexec call to another transparently) would be useful for transparent integration later.
Otherwise an intransparent modular qrexec proxy can work standalone inside a VM and achieve pretty much everything I mentioned above.

**Comment by brendanhoar** [comment]
If it has to be separated from dom0, might it be a good candidate for a low-memory mirage unikernel-based solution?

B

**Comment by 3hhh** [comment]
Hm I implemented such a qrexec proxy VM with Python asyncio and it incurred a 30-40% performance impact vs qrexec without proxy. Not great, but not too bad either.

@brendanhoar Maybe. But it's beyond my capabilities.

**Comment by marmarek** [comment]
> performance impact vs qrexec without proxy

FWIW, early qrexec implementation (AFAIR before Qubes 3?) proxied all the data through dom0. And besides significant performance penalty, it also made significant load on dom0.

While having a proxy for some calls is probably a useful feature (as you said, for example for network-based qrexec), I don't think it should be for all the calls. I can see also a value in delegating policy decision to something else (including another VM), but again - that should be for specific calls, when dom0 makes such decision (an example: some set of VMs require additional confirmation from some policy-compliance VM). I see those two features as completely independent.

On the other hand, having extra control over inter-vm communication can be achieved with other mechanisms too. Argo and XSM are some examples, but probably not the only ones.

**Comment by 3hhh** [comment]
> While having a proxy for some calls is probably a useful feature (as you said, for example for network-based qrexec), I don't think it should be for all the calls. I can see also a value in delegating policy decision to something else (including another VM), but again - that should be for specific calls, when dom0 makes such decision (an example: some set of VMs require additional confirmation from some policy-compliance VM). I see those two features as completely independent.

I agree. Actually I had decided to make the proxy intransparent so that the policy decision can still be made in dom0 via RPC policy arguments. No need to replicate code.

The use cases I currently see are:
- one-way qrexec
- defend some highly sensitive VMs against certain side channels (e.g. receiving VM blocking incoming qrexec data to communicate with the source VM in one-way qrexec)
- networked qrexec
- qrexec debugging via pcap? (not sure, might be easier to just sniff the qrexec socket directly)

I guess I'll put it somewhere on github once I have 1-2 of those implemented.

> On the other hand, having extra control over inter-vm communication can be achieved with other mechanisms too. Argo and XSM are some examples, but probably not the only ones.

True, but those would require a lot more effort.

**Comment by 3hhh** [comment]
Here's the code: https://github.com/3hhh/qubes-qrexec-proxy

**Comment by marmarek** [comment]
Kinda similar mechanism got implemented for RemoteVM (https://github.com/QubesOS/qubes-issues/issues/9015)

**Comment by ben-grande** [comment]
XSM issue as it was mentioned above: https://github.com/qubesos/qubes-issues/issues/5435
